File: tools/src/export/rosterizer.ts
/**
 * Rosterizer serializer — emits a Rosterizer-shaped roster JSON skeleton that
 * round-trips through {@link rosterizerAdapter}.
 *
 * The shape carries only fields the importer reads: `rulebook` (envelope),
 * `snapshot` (an `Asset` tree rooted at `Roster§Roster`), and per-unit
 * `item`/`name`/`quantity`/`stats.Points.value`/`assets.included`/`assets.traits`.
 * No `text`, `description`, `rules`, `lineage`, `_layers`, `classIdentity`,
 * `processed`, or `bareResourceKey` ever appear — they aren't stored in the
 * Roster and emitting them could leak prose.
 *
 * Faction and detachment display names come from {@link titleCaseId} — the
 * Roster doesn't carry the source's raw faction name, so we reconstruct it
 * from the kebab-case id. Same lossy hop as the NewRecruit JSON serializer.
 *
 * @packageDocumentation
 */
import type { Roster, RosterUnit, RosterWargear } from "../import/types.js";
import { prettyJson, titleCaseId, totalArmyPoints } from "./helpers.js";
import type { RosterSerializer } from "./serializer.js";

// Mirror the importer's constants (kept inline rather than imported so the
// exporter stays decoupled — the seams are the `item` keys themselves).
const CLS_ROSTER = "Roster";
const CLS_FACTION = "Faction";
const CLS_DETACHMENT = "Detachment";
const CLS_UNIT = "Unit";
const CLS_WEAPON = "Weapon";
const CLS_ENHANCEMENT = "Enhancement";
const CLS_BATTLE_SIZE = "Battle Size";
const CLS_TRAIT = "Trait";
const DSG_WARLORD = "Warlord";

const RULEBOOK_NAME = "40kdc";
const RULEBOOK_GAME = "Warhammer 40,000";
const RULEBOOK_PUBLISHER = "Tabletop Developer Consortium";
const RULEBOOK_URL = "https://40kdc.dev";
const RULEBOOK_GENRE = "wargame";

interface Asset {
  item: string;
  name?: string;
  quantity?: number;
  stats?: Record<string, { value: number }>;
  assets?: {
    included?: Asset[];
    traits?: Asset[];
  };
}

interface Envelope {
  slug: string;
  key: string;
  visible: "hidden" | "public" | "friends";
  locked: boolean;
  rulebook: {
    name: string;
    game: string;
    publisher: string;
    url: string;
    genre: string;
  };
  snapshot: Asset;
}

function key(classification: string, designation: string): string {
  return `${classification}§${designation}`; // §
}

function pointsStat(value: number | null | undefined): Record<string, { value: number }> | undefined {
  if (value === null || value === undefined) return undefined;
  return { Points: { value } };
}

function wargearAsset(w: RosterWargear): Asset {
  return {
    item: key(CLS_WEAPON, w.ref.raw_name),
    name: w.ref.raw_name,
    quantity: w.count,
  };
}

function enhancementAsset(u: RosterUnit): Asset | null {
  if (!u.enhancement) return null;
  return {
    item: key(CLS_ENHANCEMENT, u.enhancement.raw_name),
    name: u.enhancement.raw_name,
    quantity: 1,
    ...(u.enhancement_points !== null
      ? { stats: pointsStat(u.enhancement_points) }
      : {}),
  };
}

function warlordTraitAsset(): Asset {
  return {
    item: key(CLS_TRAIT, DSG_WARLORD),
    name: DSG_WARLORD,
    quantity: 1,
  };
}

function unitAsset(u: RosterUnit): Asset {
  const included: Asset[] = [];
  const enh = enhancementAsset(u);
  if (enh !== null) included.push(enh);
  for (const w of u.wargear) included.push(wargearAsset(w));

  const traits: Asset[] = [];
  if (u.is_warlord) traits.push(warlordTraitAsset());

  const asset: Asset = {
    item: key(CLS_UNIT, u.ref.raw_name),
    name: u.ref.raw_name,
    quantity: u.model_count,
  };
  const stats = pointsStat(u.points);
  if (stats !== undefined) asset.stats = stats;
  if (included.length > 0 || traits.length > 0) {
    asset.assets = {};
    if (included.length > 0) asset.assets.included = included;
    if (traits.length > 0) asset.assets.traits = traits;
  }
  return asset;
}

function factionAsset(roster: Roster): Asset | null {
  const display = titleCaseId(roster.faction_id);
  if (display === null) return null;
  return { item: key(CLS_FACTION, display), name: display, quantity: 1 };
}

function detachmentAsset(roster: Roster): Asset | null {
  const display = titleCaseId(roster.detachment_id);
  if (display === null) return null;
  return { item: key(CLS_DETACHMENT, display), name: display, quantity: 1 };
}

function battleSizeAsset(roster: Roster): Asset | null {
  if (roster.battle_size === "strike-force") {
    const limit = roster.points.declared_limit ?? 2000;
    const label = `Strike Force (${limit} Point limit)`;
    return { item: key(CLS_BATTLE_SIZE, label), name: label, quantity: 1 };
  }
  if (roster.battle_size === "incursion") {
    const limit = roster.points.declared_limit ?? 1000;
    const label = `Incursion (${limit} Point limit)`;
    return { item: key(CLS_BATTLE_SIZE, label), name: label, quantity: 1 };
  }
  return null;
}

export const rosterizerSerializer: RosterSerializer = {
  id: "rosterizer",

  serialize(roster: Roster): string {
    const included: Asset[] = [];
    const faction = factionAsset(roster);
    if (faction) included.push(faction);
    const detachment = detachmentAsset(roster);
    if (detachment) included.push(detachment);
    const battleSize = battleSizeAsset(roster);
    if (battleSize) included.push(battleSize);
    for (const u of roster.units) included.push(unitAsset(u));

    const total = totalArmyPoints(roster);
    const snapshot: Asset = {
      item: key(CLS_ROSTER, CLS_ROSTER),
      name: roster.name,
      quantity: 1,
      ...(total > 0 ? { stats: pointsStat(total) } : {}),
      assets: { included },
    };

    const envelope: Envelope = {
      slug: "",
      key: "",
      visible: "hidden",
      locked: false,
      rulebook: {
        name: RULEBOOK_NAME,
        game: RULEBOOK_GAME,
        publisher: RULEBOOK_PUBLISHER,
        url: RULEBOOK_URL,
        genre: RULEBOOK_GENRE,
      },
      snapshot,
    };

    return prettyJson(envelope);
  },
};

File: crates/wh40kdc/src/export/rosterizer.rs
//! Rosterizer serializer — emits a Rosterizer-shaped roster JSON skeleton
//! that round-trips through [`RosterizerAdapter`](crate::import::RosterizerAdapter).
//!
//! The shape carries only fields the importer reads: `rulebook` (envelope),
//! `snapshot` (an `Asset` tree rooted at `Roster§Roster`), and per-unit
//! `item`/`name`/`quantity`/`stats.Points.value`/`assets.included`/`assets.traits`.
//! No `text`/`description`/`rules` ever appear — they aren't stored on the
//! Roster and emitting them could leak prose.
//!
//! Faction and detachment display names come from
//! [`title_case_id`](super::helpers::title_case_id) — the Roster doesn't carry
//! the source's raw faction name, so we reconstruct it from the kebab-case
//! id. Same lossy hop as the NewRecruit JSON serializer.
//!
//! Rust mirror of `tools/src/export/rosterizer.ts`.

use serde::Serialize;

use crate::import::{BattleSize, Roster, RosterUnit, RosterWargear};

use super::helpers::{pretty_json, title_case_id, total_army_points};
use super::{ExportFormat, RosterSerializer};

const CLS_ROSTER: &str = "Roster";
const CLS_FACTION: &str = "Faction";
const CLS_DETACHMENT: &str = "Detachment";
const CLS_UNIT: &str = "Unit";
const CLS_WEAPON: &str = "Weapon";
const CLS_ENHANCEMENT: &str = "Enhancement";
const CLS_BATTLE_SIZE: &str = "Battle Size";
const CLS_TRAIT: &str = "Trait";
const DSG_WARLORD: &str = "Warlord";
const ITEM_SEPARATOR: char = '§';

const RULEBOOK_NAME: &str = "40kdc";
const RULEBOOK_GAME: &str = "Warhammer 40,000";
const RULEBOOK_PUBLISHER: &str = "Alpaca Software";
const RULEBOOK_URL: &str = "https://40kdc.dev";
const RULEBOOK_GENRE: &str = "wargame";

#[derive(Serialize)]
struct PointsStat {
    #[serde(rename = "Points")]
    points: StatValue,
}

#[derive(Serialize)]
struct StatValue {
    value: u64,
}

#[derive(Serialize, Default)]
struct AssetChildren {
    #[serde(skip_serializing_if = "Option::is_none")]
    included: Option<Vec<Asset>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    traits: Option<Vec<Asset>>,
}

#[derive(Serialize)]
struct Asset {
    item: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    quantity: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stats: Option<PointsStat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    assets: Option<AssetChildren>,
}

#[derive(Serialize)]
struct Rulebook {
    name: &'static str,
    game: &'static str,
    publisher: &'static str,
    url: &'static str,
    genre: &'static str,
}

#[derive(Serialize)]
struct Envelope {
    slug: &'static str,
    key: &'static str,
    visible: &'static str,
    locked: bool,
    rulebook: Rulebook,
    snapshot: Asset,
}

fn item_key(classification: &str, designation: &str) -> String {
    format!("{classification}{ITEM_SEPARATOR}{designation}")
}

fn points_stat(value: Option<u64>) -> Option<PointsStat> {
    value.map(|v| PointsStat {
        points: StatValue { value: v },
    })
}

fn wargear_asset(w: &RosterWargear) -> Asset {
    Asset {
        item: item_key(CLS_WEAPON, &w.ref_.raw_name),
        name: Some(w.ref_.raw_name.clone()),
        quantity: Some(w.count),
        stats: None,
        assets: None,
    }
}

fn enhancement_asset(u: &RosterUnit) -> Option<Asset> {
    let enh = u.enhancement.as_ref()?;
    Some(Asset {
        item: item_key(CLS_ENHANCEMENT, &enh.raw_name),
        name: Some(enh.raw_name.clone()),
        quantity: Some(1),
        stats: points_stat(u.enhancement_points),
        assets: None,
    })
}

fn warlord_trait_asset() -> Asset {
    Asset {
        item: item_key(CLS_TRAIT, DSG_WARLORD),
        name: Some(DSG_WARLORD.to_string()),
        quantity: Some(1),
        stats: None,
        assets: None,
    }
}

fn unit_asset(u: &RosterUnit) -> Asset {
    let mut included: Vec<Asset> = Vec::new();
    if let Some(enh) = enhancement_asset(u) {
        included.push(enh);
    }
    for w in &u.wargear {
        included.push(wargear_asset(w));
    }

    let mut traits_vec: Vec<Asset> = Vec::new();
    if u.is_warlord {
        traits_vec.push(warlord_trait_asset());
    }

    let assets = if included.is_empty() && traits_vec.is_empty() {
        None
    } else {
        Some(AssetChildren {
            included: if included.is_empty() {
                None
            } else {
                Some(included)
            },
            traits: if traits_vec.is_empty() {
                None
            } else {
                Some(traits_vec)
            },
        })
    };

    Asset {
        item: item_key(CLS_UNIT, &u.ref_.raw_name),
        name: Some(u.ref_.raw_name.clone()),
        quantity: Some(u.model_count),
        stats: points_stat(u.points),
        assets,
    }
}

fn faction_asset(roster: &Roster) -> Option<Asset> {
    let display = title_case_id(roster.faction_id.as_deref())?;
    Some(Asset {
        item: item_key(CLS_FACTION, &display),
        name: Some(display),
        quantity: Some(1),
        stats: None,
        assets: None,
    })
}

fn detachment_asset(roster: &Roster) -> Option<Asset> {
    let display = title_case_id(roster.detachment_id.as_deref())?;
    Some(Asset {
        item: item_key(CLS_DETACHMENT, &display),
        name: Some(display),
        quantity: Some(1),
        stats: None,
        assets: None,
    })
}

fn battle_size_asset(roster: &Roster) -> Option<Asset> {
    let label = match roster.battle_size? {
        BattleSize::StrikeForce => format!(
            "Strike Force ({} Point limit)",
            roster.points.declared_limit.unwrap_or(2000)
        ),
        BattleSize::Incursion => format!(
            "Incursion ({} Point limit)",
            roster.points.declared_limit.unwrap_or(1000)
        ),
    };
    Some(Asset {
        item: item_key(CLS_BATTLE_SIZE, &label),
        name: Some(label),
        quantity: Some(1),
        stats: None,
        assets: None,
    })
}

pub struct RosterizerSerializer;

impl RosterSerializer for RosterizerSerializer {
    fn id(&self) -> ExportFormat {
        ExportFormat::Rosterizer
    }

    fn serialize(&self, roster: &Roster) -> String {
        let mut included: Vec<Asset> = Vec::new();
        if let Some(f) = faction_asset(roster) {
            included.push(f);
        }
        if let Some(d) = detachment_asset(roster) {
            included.push(d);
        }
        if let Some(b) = battle_size_asset(roster) {
            included.push(b);
        }
        for u in &roster.units {
            included.push(unit_asset(u));
        }

        let total = total_army_points(roster);
        let snapshot = Asset {
            item: item_key(CLS_ROSTER, CLS_ROSTER),
            name: Some(roster.name.clone()),
            quantity: Some(1),
            stats: if total > 0 {
                points_stat(Some(total))
            } else {
                None
            },
            assets: Some(AssetChildren {
                included: Some(included),
                traits: None,
            }),
        };

        let envelope = Envelope {
            slug: "",
            key: "",
            visible: "hidden",
            locked: false,
            rulebook: Rulebook {
                name: RULEBOOK_NAME,
                game: RULEBOOK_GAME,
                publisher: RULEBOOK_PUBLISHER,
                url: RULEBOOK_URL,
                genre: RULEBOOK_GENRE,
            },
            snapshot,
        };

        pretty_json(&envelope)
    }
}
